Gems/Atom/RHI/Code/Source/RHI/MultiDeviceQueryPool.cpp

```diff
@@ -0,0 +1,249 @@
+                    });
+
+                return resultCode;
+            });
+
+        if (resultCode != ResultCode::Success)
+        {
+            // Reset already initialized device-specific QueryPools and set deviceMask to 0
+            m_deviceQueryPools.clear();
+            MultiDeviceObject::Init(static_cast<MultiDevice::DeviceMask>(0u));
+        }
+
+        return resultCode;
+    }
+
+    ResultCode MultiDeviceQueryPool::InitQuery(MultiDeviceQuery* query)
+    {
+        return InitQuery(&query, 1);
+    }
+
+    ResultCode MultiDeviceQueryPool::InitQuery(MultiDeviceQuery** queries, uint32_t queryCount)
+    {
+        AZ_Assert(queries, "Null queries");
+        auto resultCode{ ResultCode::Success };
+
+        for (auto& [deviceIndex, deviceQueryPool] : m_deviceQueryPools)
+        {
+            AZStd::vector<RHI::Ptr<Query>> deviceQueries(queryCount);
+            AZStd::vector<Query*> rawDeviceQueries(queryCount);
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                deviceQueries[index] = RHI::Factory::Get().CreateQuery();
+                rawDeviceQueries[index] = deviceQueries[index].get();
+            }
+
+            resultCode = deviceQueryPool->InitQuery(rawDeviceQueries.data(), queryCount);
+
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                queries[index]->m_deviceQueries[deviceIndex] = deviceQueries[index];
+            }
+
+            if (resultCode != ResultCode::Success)
+            {
+                break;
+            }
+        }
+
+        if (resultCode == ResultCode::Success)
+        {
+            for (auto index{ 0u }; index < queryCount; ++index)
+            {
+                MultiDeviceResourcePool::InitResource(
+                    queries[index],
+                    []()
+                    {
+                        return ResultCode::Success;
+                    });
+            }
+        }
+
+        return resultCode;
+    }
+
+    ResultCode MultiDeviceQueryPool::ValidateQueries(MultiDeviceQuery** queries, uint32_t queryCount)
+    {
+        if (!queryCount)
+        {
+            AZ_Error("RHI", false, "MultiDeviceQuery count is 0");
+            return RHI::ResultCode::InvalidArgument;
+        }
+
+        for (uint32_t i = 0; i < queryCount; ++i)
+        {
+            auto* query = queries[i];
+            if (!query)
+            {
+                AZ_Error("RHI", false, "Null query %i", i);
+                return RHI::ResultCode::InvalidArgument;
+            }
+
+            if (query->GetQueryPool() != this)
+            {
+                AZ_Error("RHI", false, "MultiDeviceQuery does not belong to this pool");
+                return RHI::ResultCode::InvalidArgument;
+            }
+        }
+        return ResultCode::Success;
+    }
+
+    ResultCode MultiDeviceQueryPool::GetResults(MultiDeviceQuery* query, uint64_t* result, uint32_t resultsCount, QueryResultFlagBits flags)
```

Comment:
The user has to initialize the MultiDeviceQueryPool with a certain deviceMask, which means the number of devices should be known by the user.

Reply:
As suggested in the previous comment, it would be easier to change the API in one of the two ways mentioned, such that a user could query results from (a) specific device(s).

Reply:
This should also be resolved with the API change I just did, introducing a method to get the necessary `resultCount`.